lib/Conversion/GPUToROCM/ConvertKernelFuncToHSACO.cpp

```diff
@@ -0,0 +1,409 @@
+  result.push_back(appendedPath.c_str());
+
+  return std::move(result);
+}
+
+std::unique_ptr<llvm::Module>
+LoadBitcodeModule(const std::string &filename, llvm::LLVMContext &llvmContext) {
+
+  llvm::SMDiagnostic diagnostic;
+  std::unique_ptr<llvm::Module> bitcodeModule(
+      llvm::parseIRFile(llvm::StringRef(filename.data(), filename.size()),
+                        diagnostic, llvmContext));
+
+  if (bitcodeModule == nullptr) {
+    llvm::errs() << diagnostic.getFilename().str() << ":"
+                 << diagnostic.getLineNo() << ":" << diagnostic.getColumnNo()
+                 << ": " << diagnostic.getMessage().str();
+  }
+
+  return bitcodeModule;
+}
+
+// Links the given llvm module with the given bitcode modules.
+void LinkWithBitcodeModules(
+    llvm::Module &llvmModule,
+    const std::vector<std::string> &bitcodeModulePaths) {
+
+  llvm::Linker linker(llvmModule);
+
+  for (auto &filename : bitcodeModulePaths) {
+
+    if (!llvm::sys::fs::exists(filename)) {
+      llvm::errs() << "bitcode module is required by this MLIR module but was "
+                      "not found at "
+                   << filename;
+      return;
+    }
+
+    std::unique_ptr<llvm::Module> bitcodeModule =
+        LoadBitcodeModule(filename, llvmModule.getContext());
+
+    if (linker.linkInModule(
+            std::move(bitcodeModule), llvm::Linker::Flags::LinkOnlyNeeded,
+            [](llvm::Module &M, const llvm::StringSet<> &GVS) {
+              internalizeModule(M, [&M, &GVS](const llvm::GlobalValue &GV) {
+                return !GV.hasName() || (GVS.count(GV.getName()) == 0);
+              });
+            })) {
+      llvm::errs() << "Error linking bitcode module from " << filename;
+      return;
+    }
+  }
+}
+
+// Returns whether the module uses any ROCDL bitcode functions. This function
+// may have false positives
+bool CouldNeedDeviceBitcode(const llvm::Module &llvmModule) {
+  for (const llvm::Function &llvmFunction : llvmModule.functions()) {
+    // This is a conservative approximation
+    //  - not all such functions are in ROCm-Device-Libs.
+    if (!llvmFunction.isIntrinsic() && llvmFunction.isDeclaration()) {
+      return true;
+    }
+  }
+  return false;
+}
+
+// Links ROCm-Device-Libs into the given module if the module needs it.
+void LinkROCDLIfNecessary(llvm::Module &llvmModule,
+                          rocm::AMDGPUVersion amdgpuVersion,
+                          const std::string &rocdlDir) {
+
+  if (!CouldNeedDeviceBitcode(llvmModule)) {
+    return;
+  }
+
+  LinkWithBitcodeModules(llvmModule, GetROCDLPaths(amdgpuVersion, rocdlDir));
+}
+
+} // anonymous namespace
+
+// Emits the given module to HSA Code Object. targetMachine is an initialized
+// TargetMachine for the AMDGPU target.
+OwnedHSACO
+GpuKernelToHSACOPass::emitModuleToHSACO(llvm::Module &llvmModule,
+                                        llvm::TargetMachine &targetMachine) {
+
+  static char tempdirTemplate[] = "/tmp/amdgpu_mlir-XXXXXX";
```

Comment:
Can you use the llvm filesystem machinery for generating a temporary file? This looks non-compliant across different platforms.

Reply:
have fixed this in the new push